htsget-search/src/storage/aws.rs

```diff
@@ -0,0 +1,398 @@
+      bucket,
+      id_resolver,
+    )
+  }
+
+  fn resolve_key<K: AsRef<str> + Send>(&self, key: &K) -> Result<String> {
+    self
+      .id_resolver
+      .resolve_id(key.as_ref())
+      .ok_or_else(|| StorageError::InvalidKey(key.as_ref().to_string()))
+  }
+
+  async fn s3_presign_url<K: AsRef<str> + Send>(
+    &self,
+    key: K,
+    range: BytesRange,
+  ) -> Result<String> {
+    let response = self
+      .client
+      .get_object()
+      .bucket(&self.bucket)
+      .key(&self.resolve_key(&key)?);
+    let response = Self::apply_range(response, range);
+    Ok(
+      response
+        .presigned(
+          PresigningConfig::expires_in(Duration::from_secs(Self::PRESIGNED_REQUEST_EXPIRY))
+            .map_err(|err| StorageError::AwsError(err.to_string(), key.as_ref().to_string()))?,
+        )
+        .await
+        .map_err(|err| StorageError::AwsError(err.to_string(), key.as_ref().to_string()))?
+        .uri()
+        .to_string(),
+    )
+  }
+
+  async fn s3_head<K: AsRef<str> + Send>(&self, key: K) -> Result<HeadObjectOutput> {
+    Ok(
+      self
+        .client
+        .head_object()
+        .bucket(&self.bucket)
+        .key(&self.resolve_key(&key)?)
+        .send()
+        .await
+        .map_err(|err| StorageError::AwsError(err.to_string(), key.as_ref().to_string()))?,
+    )
+  }
+
+  /// Returns the retrieval type of the object stored with the key.
+  pub async fn get_retrieval_type<K: AsRef<str> + Send>(&self, key: K) -> Result<Retrieval> {
+    let head = self.s3_head(&self.resolve_key(&key)?).await?;
+    Ok(
+      // Default is Standard.
+      match head.storage_class.unwrap_or(StorageClass::Standard) {
+        class @ (StorageClass::DeepArchive | StorageClass::Glacier) => {
+          Self::check_restore_header(head.restore, class)
+        }
+        class @ StorageClass::IntelligentTiering => {
+          if head.archive_status.is_some() {
+            // Not sure if this check is necessary for the archived intelligent tiering classes but
+            // it shouldn't hurt.
+            Self::check_restore_header(head.restore, class)
+          } else {
+            Immediate(class)
+          }
+        }
+        class => Immediate(class),
+      },
+    )
+  }
+
+  fn check_restore_header(restore_header: Option<String>, class: StorageClass) -> Retrieval {
+    if let Some(restore) = restore_header {
+      if restore.contains("ongoing-request=\"false\"") {
+        return Immediate(class);
+      }
+    }
+    Delayed(class)
+  }
+
+  fn apply_range(builder: GetObject, range: BytesRange) -> GetObject {
+    let range: String = range.into();
+    if range.is_empty() {
+      builder
+    } else {
+      builder.range(range)
+    }
+  }
+
+  async fn get_content<K: AsRef<str> + Send>(&self, key: K, options: GetOptions) -> Result<Bytes> {
+    // It would be nice to use a ready-made type with a ByteStream that implements AsyncRead + AsyncSeek
```

Comment:
Nice catch and worth keeping it as TODO for now... we'll see if we do need to optimize this during future stress testing.